### DISCUSSION TITLE
Does the test framework work when using Godot as the media controller?

### DISCUSSION THREAD

**Discussion by bernarma** [missionpinball/missionpinball#157] [Q&A] [answered]
Quick question, does unit testing work when using the latest version of MPF with the Godot media controller? I am hopeful it does but expecting that it doesn't.

I have gotten quite deep into the coding of the game and now need to use some unit testing as it is becoming a little fragile when I make changes / refactor.

**Comment by avanwinkle** [comment, marked as answer]
The MPF testing framework does still work the same as before, but unfortunately Godot does not have any native unit tests. There are some open-source projects (GUT and gdUnit) that may be of use for testing your media side. 

But all your mode and game logic tests in MPF will still behave as expected :)

**Reply by bernarma** [reply]
Thanks @avanwinkle. I did manage to get it working as the more rules I was coding up, the more complex it was becoming which led to fragility in making changes. I have now managed to get unit tests coded up and where there are "dependencies" on Godot (there are a few), I have managed to "simulate" that behaviour in a unit test base class that will respond to events accordingly (essentially mocking the response from Godot).

The ability to unit test MPF (heavily controlling time) has been invaluable due to the time-bound nature of a lot of the rules I am implementing. Being able to essentially stop time whilst debugging has been so helpful. I would like to eventually have a more integrated test with Godot active for certain unit tests understanding that it will definitely be slower / more real-time, but I will eventually need to do that to tease out some other bugs I have encountered and fixed the hard way.

Thanks for responding to a semi-old post.
